### PR TITLE
Support for access tokens with PoP semantics

### PR DESCRIPTION
**Pull request by manojampalam**
This is to get initial feedback on interface and implementation to support acquiring PoP based access tokens. Changes have so far been tested manually with client credential flow.

### CODE REVIEW

File: autorest/adal/popkey.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+	key, err := rsa.GenerateKey(rand.Reader, 1024)
+	if err != nil {
+		return nil, err
+	}
+
+	swk.init(key)
+	return swk, nil
+}
+
+var pswKey *swKey
```
Comment (keegan-caruso) [inline autorest/adal/popkey.go:120]: do these need to be globals? Could they be held on a type that encapsulates this logic?

File: autorest/adal/token.go [hunk 2]
```diff
@@ -916,6 +922,11 @@ func (spt *ServicePrincipalToken) refreshInternal(ctx context.Context, resource
 		v := url.Values{}
 		v.Set("client_id", spt.inner.ClientID)
 		v.Set("resource", resource)
+		if spt.enablePoP == true {
```
Comment (keegan-caruso) [inline autorest/adal/token.go]: can drop ` == true` since enablePoP is a bool

File: autorest/adal/popkey.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+	// compute JWK thumbprint
+	//jwk format - e, kty, n - in lexicographic order
+	// - https://tools.ietf.org/html/rfc7638#section-3.3
+	// - https://tools.ietf.org/html/rfc7638#section-3.1
+	jwk := fmt.Sprintf(`{"e":"%s","kty":"RSA","n":"%s"}`, eB64, nB64)
+	jwkS256 := sha256.Sum256([]byte(jwk))
+	swk.jwkTP = base64.RawURLEncoding.EncodeToString(jwkS256[:])
+
+	//req_cnf - base64URL("{"kid":"jwkTP","xms_ksl":"sw"}")
```
Comment (keegan-caruso) [inline autorest/adal/popkey.go:96]: should this value be base64 encoded?

File: autorest/adal/poptoken.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	ts := time.Now().Unix()
+	nonce := uuid.New().String()
+	nonce = strings.Replace(nonce, "-", "", -1)
+	header := fmt.Sprintf(`{"typ":"pop","alg":"%s","kid":"%s"}`, tkn.poPKey.Alg(), tkn.poPKey.KeyID())
+	headerB64 := base64.RawURLEncoding.EncodeToString([]byte(header))
+	payload := fmt.Sprintf(`{"at":"%s","ts":%d,"u":"%s","cnf":{"jwk":%s},"nonce":"%s"}`, tkn.AccessToken, ts, host, tkn.poPKey.JWK(), nonce)
+	payloadB64 := base64.RawURLEncoding.EncodeToString([]byte(payload))
+	h256 := sha256.New()
+	h256.Write([]byte(headerB64 + "." + payloadB64))
+	signature, err := tkn.poPKey.Sign(h256.Sum(nil))
```
Comment (keegan-caruso) [inline autorest/adal/poptoken.go]: Can you use [Sum256](https://golang.org/pkg/crypto/sha256/#Sum256) here?

File: autorest/adal/token.go (hunk 2 above)
Comment (manojampalam) [reply autorest/adal/token.go]: Dropped

File: autorest/adal/poptoken.go (hunk 4 above)
Comment (manojampalam) [reply autorest/adal/poptoken.go]: Thanks for the tip. Used Sum256 directly.

File: autorest/adal/popkey.go (hunk 1 above)
Comment (manojampalam) [reply autorest/adal/popkey.go:120]: We could put these 2 inside a struct but would need a global instance of that struct, that requires initialization, perhaps in an init(). Found it an overkill for the need here. Is there a better option ?

Comment (keegan-caruso) [reply autorest/adal/popkey.go:120]: I see, if it has to be a global anyways, not a lot of value add. I was hoping something else could own it, but I'm not seeing much.

**Review by keegan-caruso — APPROVED** [review]

**Pull request merged**